File: back/src/main/java/com/back/domain/waiting/service/WaitingQueueManager.java
package com.back.domain.waiting.service;

import com.back.domain.waiting.dto.QueueStatusDto;
import com.back.global.exception.ErrorCode;
import com.back.global.exception.ServiceException;
import lombok.RequiredArgsConstructor;
import org.redisson.api.*;
import org.springframework.stereotype.Component;

import java.time.Duration;
import java.util.*;

/**
 * Redis 기반 대기열 상태 관리 컴포넌트.
 *
 * <p>Redisson 클라이언트를 통해 대기(Wait) / 진입(Active) 큐를 관리합니다.
 * <ul>
 *   <li>대기열 ({@code RScoredSortedSet}): 유저 ID를 Score(발급 순번) 기준으로 정렬</li>
 *   <li>진입열 ({@code RScoredSortedSet}): 유저 ID를 Score(만료 타임스탬프) 기준으로 정렬</li>
 *   <li>진입 토큰 ({@code RBucket}): 유저별 UUID 토큰, TTL 10분</li>
 *   <li>활성 회차 목록 ({@code RSet}): 스케줄러 루프 최소화용</li>
 * </ul>
 */
@Component
@RequiredArgsConstructor
public class WaitingQueueManager {

    private final RedissonClient redissonClient;

    private static final String WAIT_KEY_PREFIX = "queue:wait:schedule:";
    private static final String SEQUENCE_KEY_PREFIX = "queue:wait:sequence:schedule:";
    private static final String ACTIVE_TOKEN_KEY_PREFIX = "queue:active:token:";
    private static final String ACTIVE_SCHEDULES_KEY = "queue:active:schedules";

    // ─── 대기열 등록 ──────────────────────────────────────────────────────────

    /**
     * 대기열에 유저를 등록합니다. 이미 등록된 경우 기존 순번을 반환합니다.
     *
     * @return 현재 대기 순번 (1-based)
     */
    public Long registerWaiting(Long scheduleId, Long userId) {
        String waitKey = generateWaitKey(scheduleId);
        String seqKey = generateSequenceKey(scheduleId);
        String user = userId.toString();

        // 활성 회차 목록에 추가
        RSet<String> activeSchedules = redissonClient.getSet(ACTIVE_SCHEDULES_KEY);
        activeSchedules.add(scheduleId.toString());

        // Lua Script: 이미 등록된 경우 기존 순번 반환, 신규면 INCR 후 ZADD
        RScript script = redissonClient.getScript();
        Long rank = script.eval(
                RScript.Mode.READ_WRITE,
                """
                local exists = redis.call('ZSCORE', KEYS[1], ARGV[1])
                if not exists then
                  local sequence = redis.call('INCR', KEYS[2])
                  redis.call('ZADD', KEYS[1], sequence, ARGV[1])
                end
                local rank = redis.call('ZRANK', KEYS[1], ARGV[1])
                if not rank then
                  return -1
                end
                return rank + 1
                """,
                RScript.ReturnType.INTEGER,
                Arrays.asList(waitKey, seqKey),
                user
        );

        if (rank == null || rank < 1) {
            throw new ServiceException(ErrorCode.WAITING_QUEUE_REGISTER_FAILED);
        }
        return rank;
    }

    // ─── 대기 순번 조회 ───────────────────────────────────────────────────────

    public Long showWaitingRank(Long scheduleId, Long userId) {
        RScoredSortedSet<String> waitSet = redissonClient.getScoredSortedSet(generateWaitKey(scheduleId));
        Integer rank = waitSet.rank(userId.toString());
        if (rank == null) {
            throw new ServiceException(ErrorCode.WAITING_QUEUE_NOT_FOUND);
        }
        return rank + 1L;
    }

    // ─── 대기열 취소 ──────────────────────────────────────────────────────────

    public boolean cancelWaiting(Long scheduleId, Long userId) {
        RScoredSortedSet<String> waitSet = redissonClient.getScoredSortedSet(generateWaitKey(scheduleId));
        return waitSet.remove(userId.toString());
    }

    public boolean cancelActiveUser(Long scheduleId, Long userId) {
        RScoredSortedSet<String> activeSet = redissonClient.getScoredSortedSet(generateQueueActiveKey(scheduleId));
        boolean removed = activeSet.remove(userId.toString());
        redissonClient.getBucket(generateActiveTokenKey(scheduleId, userId)).delete();
        return removed;
    }

    public void removeActiveUser(Long scheduleId, Long userId) {
        RScoredSortedSet<String> activeSet = redissonClient.getScoredSortedSet(generateQueueActiveKey(scheduleId));
        activeSet.remove(userId.toString());
        redissonClient.getBucket(generateActiveTokenKey(scheduleId, userId)).delete();
    }

    // ─── 만료된 진입자 청소 ───────────────────────────────────────────────────

    public long removeExpiredActiveUsers(Long scheduleId) {
        RScoredSortedSet<String> activeSet = redissonClient.getScoredSortedSet(generateQueueActiveKey(scheduleId));
        // Score가 현재 시간(ms)보다 작은 항목 = 만료된 유저
        int removed = activeSet.removeRangeByScore(0, true, System.currentTimeMillis(), true);
        return removed;
    }

    // ─── 진입 허가 ───────────────────────────────────────────────────────────

    /**
     * 대기열 앞쪽 유저를 진입열로 이동시킵니다.
     *
     * @param capacity  현재 진입 가능한 최대 인원
     * @param batchSize 한 번에 허가할 최대 인원
     * @param ttl       진입 토큰 유효 시간
     * @return 진입 허가된 userId 목록
     */
    public List<Long> addActiveUser(Long scheduleId, long capacity, int batchSize, Duration ttl) {
        String waitKey = generateWaitKey(scheduleId);
        String activeKey = generateQueueActiveKey(scheduleId);
        long now = System.currentTimeMillis();
        long expiredAt = now + ttl.toMillis();

        RScript script = redissonClient.getScript();
        List<String> userIds = script.eval(
                RScript.Mode.READ_WRITE,
                """
                redis.call('ZREMRANGEBYSCORE', KEYS[2], '-inf', ARGV[3])
                local activeCount = redis.call('ZCARD', KEYS[2])
                local availableSlots = tonumber(ARGV[1]) - activeCount
                if availableSlots <= 0 then
                  return {}
                end
                local popCount = math.min(availableSlots, tonumber(ARGV[2]))
                local users = redis.call('ZRANGE', KEYS[1], 0, popCount - 1)
                if #users == 0 then
                  return {}
                end
                redis.call('ZREM', KEYS[1], unpack(users))
                for _, u in ipairs(users) do
                  redis.call('ZADD', KEYS[2], ARGV[4], u)
                end
                return users
                """,
                RScript.ReturnType.MULTI,
                Arrays.asList(waitKey, activeKey),
                String.valueOf(capacity),
                String.valueOf(batchSize),
                String.valueOf(now),
                String.valueOf(expiredAt)
        );

        if (userIds == null || userIds.isEmpty()) {
            return List.of();
        }
        return userIds.stream().map(Long::valueOf).toList();
    }

    // ─── 토큰 발행 및 검증 ────────────────────────────────────────────────────

    public String issueToken(Long scheduleId, Long userId, Duration ttl) {
        String entryToken = UUID.randomUUID().toString();
        RBucket<String> bucket = redissonClient.getBucket(generateActiveTokenKey(scheduleId, userId));
        bucket.set(entryToken, ttl);
        return entryToken;
    }

    public String getActiveToken(Long scheduleId, Long userId) {
        RScoredSortedSet<String> activeSet = redissonClient.getScoredSortedSet(generateQueueActiveKey(scheduleId));
        Double score = activeSet.getScore(userId.toString());
        if (score == null || score <= System.currentTimeMillis()) {
            return null;
        }
        RBucket<String> bucket = redissonClient.getBucket(generateActiveTokenKey(scheduleId, userId));
        return bucket.get();
    }

    // ─── 큐 상태 조회 ────────────────────────────────────────────────────────

    public QueueStatusDto getQueueStatus(Long scheduleId) {
        RScoredSortedSet<String> waitSet = redissonClient.getScoredSortedSet(generateWaitKey(scheduleId));
        long totalWaitingCount = waitSet.size();
        if (totalWaitingCount == 0) {
            return new QueueStatusDto(0L, 0L);
        }

        // 대기열의 가장 첫 번째 항목(가장 작은 Score)의 Score에서 1을 빼면 이전 허가 번호
        Double firstScore = waitSet.firstScore();
        long currentAllowedSequence = (firstScore != null) ? firstScore.longValue() - 1 : 0L;
        return new QueueStatusDto(currentAllowedSequence, totalWaitingCount);
    }

    public Long getQueueSequence(Long scheduleId, Long userId) {
        RScoredSortedSet<String> waitSet = redissonClient.getScoredSortedSet(generateWaitKey(scheduleId));
        Double score = waitSet.getScore(userId.toString());
        return (score != null) ? score.longValue() : 0L;
    }

    // ─── 큐 비어있음 확인 ────────────────────────────────────────────────────

    public boolean isQueueEmpty(Long scheduleId) {
        RScoredSortedSet<String> waitSet = redissonClient.getScoredSortedSet(generateWaitKey(scheduleId));
        RScoredSortedSet<String> activeSet = redissonClient.getScoredSortedSet(generateQueueActiveKey(scheduleId));
        return waitSet.isEmpty() && activeSet.isEmpty();
    }

    // ─── 전체 대기열 삭제 ────────────────────────────────────────────────────

    public void clearWaitingQueue(Long scheduleId) {
        redissonClient.getScoredSortedSet(generateWaitKey(scheduleId)).delete();
    }

    // ─── 활성 회차 목록 관리 ──────────────────────────────────────────────────

    public Set<String> getActiveScheduleIds() {
        RSet<String> activeSchedules = redissonClient.getSet(ACTIVE_SCHEDULES_KEY);
        return activeSchedules.readAll();
    }

    public void removeFromActiveSchedules(String scheduleIdStr) {
        RSet<String> activeSchedules = redissonClient.getSet(ACTIVE_SCHEDULES_KEY);
        activeSchedules.remove(scheduleIdStr);
    }

    // ─── 키 생성 유틸 ────────────────────────────────────────────────────────

    private String generateWaitKey(Long scheduleId) {
        return WAIT_KEY_PREFIX + scheduleId;
    }

    private String generateSequenceKey(Long scheduleId) {
        return SEQUENCE_KEY_PREFIX + scheduleId;
    }

    public static String generateActiveTokenKey(Long scheduleId, Long userId) {
        return ACTIVE_TOKEN_KEY_PREFIX + scheduleId + ":" + userId;
    }

    public static String generateQueueActiveKey(Long scheduleId) {
        return "queue:active:schedule:%d".formatted(scheduleId);
    }
}

File: back/src/main/kotlin/com/back/domain/waiting/service/WaitingQueueManager.kt
package com.back.domain.waiting.service

import com.back.domain.waiting.dto.QueueStatusDto
import com.back.global.exception.ErrorCode
import com.back.global.exception.ServiceException
import org.springframework.data.redis.core.StringRedisTemplate
import org.springframework.data.redis.core.script.DefaultRedisScript
import org.springframework.data.redis.core.script.RedisScript
import org.springframework.stereotype.Component
import java.time.Duration
import java.util.UUID

@Component
class WaitingQueueManager(
    private val stringRedisTemplate: StringRedisTemplate
) {

    fun registerWaiting(scheduleId: Long, userId: Long): Long {
        val waitKey = generateWaitKey(scheduleId)
        val seqKey = generateSequenceKey(scheduleId)
        val user = userId.toString()

        stringRedisTemplate.opsForSet().add(ACTIVE_SCHEDULES_KEY, scheduleId.toString())

        val rank: Long? = stringRedisTemplate.execute(
            REGISTER_WAIT_SCRIPT,
            listOf(waitKey, seqKey),
            user
        )

        if (rank == null || rank < 1) {
            throw ServiceException(ErrorCode.WAITING_QUEUE_REGISTER_FAILED)
        }
        return rank
    }

    fun showWaitingRank(scheduleId: Long, userId: Long): Long =
        findWaitingRank(scheduleId, userId)
            ?: throw ServiceException(ErrorCode.WAITING_QUEUE_NOT_FOUND)

    fun findWaitingRank(scheduleId: Long, userId: Long): Long? =
        stringRedisTemplate.opsForZSet()
            .rank(generateWaitKey(scheduleId), userId.toString())
            ?.let { it + 1L }

    fun cancelWaiting(scheduleId: Long, userId: Long): Boolean =
        (stringRedisTemplate.opsForZSet().remove(generateWaitKey(scheduleId), userId.toString()) ?: 0L) > 0L

    fun cancelActiveUser(scheduleId: Long, userId: Long): Boolean =
        deleteActiveUser(scheduleId, userId)

    fun removeActiveUser(scheduleId: Long, userId: Long) {
        deleteActiveUser(scheduleId, userId)
    }

    private fun deleteActiveUser(scheduleId: Long, userId: Long): Boolean {
        val removed = (stringRedisTemplate.opsForZSet().remove(generateQueueActiveKey(scheduleId), userId.toString()) ?: 0L) > 0L
        stringRedisTemplate.delete(generateActiveTokenKey(scheduleId, userId))
        return removed
    }

    fun hasValidSession(scheduleId: Long, userId: Long): Boolean {
        val expiresAt = stringRedisTemplate.opsForZSet()
            .score(generateQueueActiveKey(scheduleId), userId.toString()) ?: return false
        return expiresAt >= System.currentTimeMillis()
    }

    fun getStoredToken(scheduleId: Long, userId: Long): String? =
        stringRedisTemplate.opsForValue().get(generateActiveTokenKey(scheduleId, userId))

    fun removeExpiredActiveUsers(scheduleId: Long): Long {
        val removed = stringRedisTemplate.opsForZSet()
            .removeRangeByScore(generateQueueActiveKey(scheduleId), 0.0, System.currentTimeMillis().toDouble()) ?: 0L
        return removed
    }

    fun addActiveUser(scheduleId: Long, capacity: Long, batchSize: Int, ttl: Duration): List<QueueAdmission> {
        val waitKey = generateWaitKey(scheduleId)
        val activeKey = generateQueueActiveKey(scheduleId)
        val now = System.currentTimeMillis()
        val expiredAt = now + ttl.toMillis()
        val tokenCandidates = List(batchSize) { UUID.randomUUID().toString() }

        val rawResult: Any? = stringRedisTemplate.execute(
            ADD_ACTIVE_USER_SCRIPT,
            listOf(waitKey, activeKey),
            capacity.toString(),
            batchSize.toString(),
            now.toString(),
            expiredAt.toString(),
            ttl.toMillis().toString(),
            generateActiveTokenKeyPrefix(scheduleId),
            *tokenCandidates.toTypedArray(),
        )

        val values = rawResult as? List<*> ?: return emptyList()
        return values.chunked(2).mapNotNull { admission ->
            val userId = admission.getOrNull(0)?.toString()?.toLongOrNull() ?: return@mapNotNull null
            val entryToken = admission.getOrNull(1)?.toString() ?: return@mapNotNull null
            QueueAdmission(userId, entryToken, expiredAt)
        }
    }

    fun getActiveToken(scheduleId: Long, userId: Long): String? {
        val score = stringRedisTemplate.opsForZSet().score(generateQueueActiveKey(scheduleId), userId.toString())
        if (score == null || score <= System.currentTimeMillis()) {
            return null
        }
        return stringRedisTemplate.opsForValue().get(generateActiveTokenKey(scheduleId, userId))
    }

    fun getConnectionSnapshot(scheduleId: Long, userId: Long): QueueConnectionSnapshot {
        val result = stringRedisTemplate.execute(
            GET_CONNECTION_SNAPSHOT_SCRIPT,
            listOf(
                generateQueueActiveKey(scheduleId),
                generateActiveTokenKey(scheduleId, userId),
                generateWaitKey(scheduleId),
            ),
            userId.toString(),
            System.currentTimeMillis().toString(),
        ) ?: throw ServiceException(ErrorCode.WAITING_QUEUE_REGISTER_FAILED)

        return when (result.firstOrNull()?.toString()) {
            "ACTIVE" -> result.getOrNull(1)?.toString()
                ?.let(QueueConnectionSnapshot::Active)
                ?: throw ServiceException(ErrorCode.WAITING_QUEUE_REGISTER_FAILED)

            "WAITING" -> {
                val rank = result.getOrNull(1)?.toString()?.toLongOrNull()
                val queueNumber = result.getOrNull(2)?.toString()?.toLongOrNull()
                if (rank == null || queueNumber == null) {
                    throw ServiceException(ErrorCode.WAITING_QUEUE_REGISTER_FAILED)
                }
                QueueConnectionSnapshot.Waiting(rank, queueNumber)
            }

            "NOT_REGISTERED" -> QueueConnectionSnapshot.NotRegistered
            else -> throw ServiceException(ErrorCode.WAITING_QUEUE_REGISTER_FAILED)
        }
    }

    fun getQueueStatus(scheduleId: Long): QueueStatusDto {
        val waitKey = generateWaitKey(scheduleId)
        val totalWaitingCount = stringRedisTemplate.opsForZSet().size(waitKey) ?: 0L
        if (totalWaitingCount == 0L) {
            return QueueStatusDto(0L, 0L)
        }

        val rangeWithScores = stringRedisTemplate.opsForZSet().rangeWithScores(waitKey, 0, 0)
        val firstScore = rangeWithScores?.firstOrNull()?.score
        val currentAllowedSequence = if (firstScore != null) firstScore.toLong() - 1 else 0L
        return QueueStatusDto(currentAllowedSequence, totalWaitingCount)
    }

    fun getQueueSequence(scheduleId: Long, userId: Long): Long {
        val score = stringRedisTemplate.opsForZSet().score(generateWaitKey(scheduleId), userId.toString())
        return score?.toLong() ?: 0L
    }

    fun isQueueEmpty(scheduleId: Long): Boolean {
        val waitSize = stringRedisTemplate.opsForZSet().size(generateWaitKey(scheduleId)) ?: 0L
        val activeSize = stringRedisTemplate.opsForZSet().size(generateQueueActiveKey(scheduleId)) ?: 0L
        return waitSize == 0L && activeSize == 0L
    }

    fun hasWaitingUsers(scheduleId: Long): Boolean =
        (stringRedisTemplate.opsForZSet().size(generateWaitKey(scheduleId)) ?: 0L) > 0L

    fun clearWaitingQueue(scheduleId: Long) {
        stringRedisTemplate.delete(generateWaitKey(scheduleId))
    }

    fun getActiveScheduleIds(): Set<String> {
        return stringRedisTemplate.opsForSet().members(ACTIVE_SCHEDULES_KEY) ?: emptySet()
    }

    fun removeFromActiveSchedules(scheduleIdStr: String) {
        stringRedisTemplate.opsForSet().remove(ACTIVE_SCHEDULES_KEY, scheduleIdStr)
    }

    private fun generateWaitKey(scheduleId: Long): String = "$WAIT_KEY_PREFIX$scheduleId"
    private fun generateSequenceKey(scheduleId: Long): String = "$SEQUENCE_KEY_PREFIX$scheduleId"

    companion object {
        private const val WAIT_KEY_PREFIX = "queue:wait:schedule:"
        private const val SEQUENCE_KEY_PREFIX = "queue:wait:sequence:schedule:"
        private const val ACTIVE_TOKEN_KEY_PREFIX = "queue:active:token:"
        private const val ACTIVE_SCHEDULES_KEY = "queue:active:schedules"

        private val REGISTER_WAIT_LUA = """
            local exists = redis.call('ZSCORE', KEYS[1], ARGV[1])
            if not exists then
              local sequence = redis.call('INCR', KEYS[2])
              redis.call('ZADD', KEYS[1], sequence, ARGV[1])
            end
            local rank = redis.call('ZRANK', KEYS[1], ARGV[1])
            if not rank then
              return -1
            end
            return rank + 1
        """.trimIndent()

        private val ADD_ACTIVE_USER_LUA = """
            redis.call('ZREMRANGEBYSCORE', KEYS[2], '-inf', tonumber(ARGV[3]))
            local activeCount = redis.call('ZCARD', KEYS[2])
            local availableSlots = tonumber(ARGV[1]) - activeCount
            if availableSlots <= 0 then
              return {}
            end
            local popCount = math.min(availableSlots, tonumber(ARGV[2]))
            local users = redis.call('ZRANGE', KEYS[1], 0, popCount - 1)
            if #users == 0 then
              return {}
            end
            redis.call('ZREM', KEYS[1], unpack(users))
            local result = {}
            for i, u in ipairs(users) do
              local token = ARGV[6 + i]
              redis.call('ZADD', KEYS[2], tonumber(ARGV[4]), u)
              redis.call('SET', ARGV[6] .. u, token, 'PX', ARGV[5])
              table.insert(result, u)
              table.insert(result, token)
            end
            return result
        """.trimIndent()

        private val GET_CONNECTION_SNAPSHOT_LUA = """
            local expiresAt = redis.call('ZSCORE', KEYS[1], ARGV[1])
            if expiresAt then
              if tonumber(expiresAt) > tonumber(ARGV[2]) then
                local token = redis.call('GET', KEYS[2])
                if token then
                  return {'ACTIVE', token}
                end
                redis.call('ZREM', KEYS[1], ARGV[1])
              else
                redis.call('ZREM', KEYS[1], ARGV[1])
                redis.call('DEL', KEYS[2])
              end
            end

            local rank = redis.call('ZRANK', KEYS[3], ARGV[1])
            if rank then
              local queueNumber = redis.call('ZSCORE', KEYS[3], ARGV[1])
              return {'WAITING', tostring(rank + 1), tostring(queueNumber)}
            end
            return {'NOT_REGISTERED'}
        """.trimIndent()

        private val REGISTER_WAIT_SCRIPT: RedisScript<Long> = DefaultRedisScript(REGISTER_WAIT_LUA, Long::class.java)

        private val ADD_ACTIVE_USER_SCRIPT: RedisScript<List<*>> =
            DefaultRedisScript(ADD_ACTIVE_USER_LUA, List::class.java)

        private val GET_CONNECTION_SNAPSHOT_SCRIPT: RedisScript<List<*>> =
            DefaultRedisScript(GET_CONNECTION_SNAPSHOT_LUA, List::class.java)

        private fun generateActiveTokenKeyPrefix(scheduleId: Long): String =
            "$ACTIVE_TOKEN_KEY_PREFIX$scheduleId:"

        @JvmStatic
        fun generateActiveTokenKey(scheduleId: Long, userId: Long): String =
            "$ACTIVE_TOKEN_KEY_PREFIX$scheduleId:$userId"

        @JvmStatic
        fun generateQueueActiveKey(scheduleId: Long): String =
            "queue:active:schedule:$scheduleId"
    }
}
